### PR TITLE
push: send a module as a manifest and blobs

### PR DESCRIPTION
**Pull request by twilly**
Let's push the module to the registry as plain files represented as a bag of blobs and a manifest.

This does not remove pushing a protobuf encoded module.

Additionally, the change to deprecate the module field in a push request makes testing for the field a linting issue. Let's continue testing this field as it's still in use.

### CODE REVIEW

File: private/pkg/manifest/module.go [hunk 1]
```diff
@@ -34,3 +36,16 @@ func NewDigestFromBlobHash(hash *modulev1alpha1.Hash) (*Digest, error) {
...
+// NewBlobFromBytes creates a module Blob from in-memory content.
+func NewBlobFromBytes(content []byte) *modulev1alpha1.Blob {
+	blob := &modulev1alpha1.Blob{
+		Hash: &modulev1alpha1.Hash{
+			Kind:   modulev1alpha1.HashKind_HASH_KIND_SHAKE256,
+			Digest: make([]byte, 64),
+		},
+		Content: content,
+	}
+	sha3.ShakeSum256(blob.Hash.Digest, content)
```
Comment (unmultimedio) [inline private/pkg/manifest/module.go:49]: Why not using the digester interface? We could just receive the Digest+Content and optionally validate?

File: private/buf/cmd/buf/command/push/push_test.go [hunk 2]
```diff
@@ -143,6 +146,29 @@ func TestPush(t *testing.T) {
...
+		},
+	)
+	require.NoError(t, err)
+	blobs, err := bucketBlobs(context.Background(), bucket)
+	require.NoError(t, err)
+	assert.Equal(t, 2, len(blobs))
+	digests := make(map[string]struct{})
+	for _, blob := range blobs {
+		shake := modulev1alpha1.HashKind_HASH_KIND_SHAKE256
+		assert.Equal(t, blob.Hash.Kind, shake)
```
[the remark this answers is not in the document]
Comment (twilly) [reply private/buf/cmd/buf/command/push/push_test.go:165]: The line got too long for my tastes. Instead I'll break the args into multiple lines.

File: private/pkg/manifest/module.go (hunk 1 above)
Comment (twilly) [reply private/pkg/manifest/module.go:49]: We can now! I wrote this before but I missed the new stuff when rebasing. I'll clean this up. :)

File: private/pkg/manifest/module.go [hunk 3]
```diff
@@ -34,3 +37,26 @@ func NewDigestFromBlobHash(hash *modulev1alpha1.Hash) (*Digest, error) {
...
+
+// NewBlobFromReader creates a module Blob from content, which is read until
+// completion. The returned blob contains all bytes read.
+func NewBlobFromReader(content io.Reader) (*modulev1alpha1.Blob, error) {
+	digester, err := NewDigester(DigestTypeShake256)
+	if err != nil {
+		return nil, err
+	}
+	var contentInMemory bytes.Buffer
+	tee := io.TeeReader(content, &contentInMemory)
```
[the remark this answers is not in the document]
Comment (twilly) [reply private/pkg/manifest/module.go:49]: It's one of my favorite io tools. :)

File: private/buf/cmd/buf/command/push/push.go [hunk 4]
```diff
@@ -204,3 +217,49 @@ func run(
...
+		return nil, err
+	}
+	return manifest.NewBlobFromReader(bytes.NewReader(manifestText))
+}
+
+// bucketBlobs returns a deduplicated set of blobs for the files in bucket.
+func bucketBlobs(
+	ctx context.Context,
+	bucket storage.ReadBucket,
+) ([]*modulev1alpha1.Blob, error) {
```
[the remark this answers is not in the document]
Comment (twilly) [reply private/buf/cmd/buf/command/push/push.go:241]: That could be useful to deduplicate a bit of code here. I'll make a cleanup ticket for this when our pair of PRs land.

[the remark this answers is not in the document]
Comment (twilly) [reply private/buf/cmd/buf/command/push/push.go:241]: https://linear.app/bufbuild/issue/TCN-746

File: private/buf/cmd/buf/command/push/push.go [hunk 5]
```diff
@@ -204,3 +217,49 @@ func run(
...
+	var blobs []*modulev1alpha1.Blob
+	haveBlob := make(map[string]struct{})
+	if walkErr := bucket.Walk(ctx, "", func(info storage.ObjectInfo) error {
```
Comment (unmultimedio) [inline private/buf/cmd/buf/command/push/push.go:244]: Just tested this, it's including everything in the `buf.yaml` directory. We should have an allow-list when pushing blobs, only "pushable" config files and protos? Here's what I got in the pushed manifest blob (along with all the related blobs):

```
shake256:3a44e5c70e36  LICENSE
shake256:221e20f240f8  buf.gen.yaml
shake256:4ee086005bd8  buf.lock
shake256:09bd2aa76a9d  buf.md
shake256:dd042e390dfd  buf.yaml
shake256:e7f2b8af2efd  foopkg/v1/1.proto
shake256:a9c48acdec2a  foopkg/v1/2.proto
shake256:c9e4511deaf2  foopkg/v1/deps.proto
shake256:b1cff4fd0ef9  some.other.file
```

Pretty sure `buf.gen.yaml` and `some.other.file` shouldn't have been pushed.

**Pull request merged**